net: tcp: Send RST reply for unexpected TCP packets

rlubos opened the pull request:
This PR improves the handling of unexpected packets within Zephyr's TCP stack. The TCP stack will now send RST packet in a response to unexpected TCP packets, to address issues mostly related to ungraceful TCP connection shutdown (for example due to reboot):
1) Invalid ACK value received during handshake (connection still open
   on the peer side),
2) Unexpected data packet on a listening port (accepted connection
   closed),
3) SYN received on a closed port.

For this, a new helper function was introduced which allows to send an RST packet to **any** destination (current functions allow only to send TCP packets to destinations associated with a TCP connection context, which is not always the case in this scenario).

Finally, add a few new tests to cover this functionality.

Fixes #63966

rlubos committed:
net: tcp: Add helper function to send RST packet w/o active connection

Add a helper function which allows to send a RST packet in response to
an unexpected TCP packet, w/o associated connection or net context.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: tcp: Send RST reply for unexpected TCP packets

Send RST as a reply for unexpected TCP packets in the following
scenarios:
1) Unexpected ACK value received during handshake (connection still open
   on the peer side),
2) Unexpected data packet on a listening port (accepted connection
   closed),
3) SYN received on a closed port.

This allows the other end to detect that the connection is no longer
valid (for example due to reboot) and release the resources.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
tests: net: tcp: Add tests to cover improved RST handling

Add tests which verify that Zephyr's TCP stack replies with RST packet
as expected.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

jukkar commented on subsys/net/ip/connection.c:
```diff
@@ -828,11 +828,12 @@ enum net_verdict net_conn_input(struct net_pkt *pkt,
-		conn_send_icmp_error(pkt);
-
 		if (IS_ENABLED(CONFIG_NET_TCP) && proto == IPPROTO_TCP) {
+			net_tcp_reply_rst(pkt);
```
In Linux one can use netfilter to prevent RST to be sent out, and in Zephyr that might be possible at some point (currently not possible). So I was thinking that should we have a Kconfig option to allow earlier behavior, so that RST would not be sent if option is set, by default the option could be off so RST would be sent out. WDYT?

rlubos replied:
@jukkar I've added a Kconfig option to configure whether to send RST packet on closed port. If disabled, ICMP is sent instead as it used to.

rlubos committed:
net: tcp: Add Kconfig option to enable TCP RST on unbound ports

Add Kconfig option to control TCP RST behavior on connection attempts on
unbound ports. If enabled, TCP stack will reply with RST packet (enabled
by default).

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
tests: lib: thrift: Fix timing issues with TCP

The test delegates putting the server socket into listening (i. e.
calling listen() on the socket) to a separate thread, which did have a
chance to run before client attempted to establish TCP connection.

This was not visible before, as we did not reply with RST to a
connection attempt on a closed port, so the connection was eventually
establish after SYN retransmission. But as we do reject such a
connection now with RST, the connection attempt failed. Therefore, a
small delay was added after spawning the server thread, to give it a
chance to configure the server socket.

Additionally, lower the CONFIG_NET_TCP_TIME_WAIT_DELAY value so that TCP
contexts are released earlier, and add a respective delay in the test
teardown function. Not doing so also triggered unneeded SYN
retransmissions, as there were no enough TCP context to accept the
incoming connection, before freeing the resources allocated for the
previous one.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
@cfriedt Thrift tests needed small adjustments due to timing issues, please check the commit message for details.
